chain/network/src/routing/graph/mod.rs

```diff
@@ -0,0 +1,315 @@
+        if peer_id == self.config.node_id || self.peer_reachable_at.contains_key(&peer_id) {
+            return;
+        }
+        let edges = match self.store.pop_component(&peer_id) {
+            Ok(edges) => edges,
+            Err(e) => {
+                tracing::warn!("self.store.pop_component({}): {}", peer_id, e);
+                return;
+            }
+        };
+        for e in edges {
+            self.update_edge(now, e);
+        }
+    }
+
+    /// Prunes peers unreachable since <unreachable_since> (and their adjacent edges)
+    /// from the in-mem graph and stores them in DB.
+    fn prune_unreachable_peers(&mut self, unreachable_since: time::Instant) {
+        // Select peers to prune.
+        let mut peers = HashSet::new();
+        for k in self.edges.keys() {
+            for peer_id in [&k.0, &k.1] {
+                if self
+                    .peer_reachable_at
+                    .get(peer_id)
+                    .map(|t| t < &unreachable_since)
+                    .unwrap_or(true)
+                {
+                    peers.insert(peer_id.clone());
+                }
+            }
+        }
+        if peers.is_empty() {
+            return;
+        }
+
+        // Prune peers from peer_reachable_at.
+        for peer_id in &peers {
+            self.peer_reachable_at.remove(&peer_id);
+        }
+
+        // Prune edges from graph.
+        let edges = self.remove_adjacent_edges(&peers);
+
+        // Store the pruned data in DB.
+        if let Err(e) = self.store.push_component(&peers, &edges) {
+            tracing::warn!("self.store.push_component(): {}", e);
+        }
+    }
+
+    /// 1. Adds edges to the graph.
+    /// 2. Prunes expired edges.
+    /// 3. Prunes unreachable graph components.
+    /// 4. Recomputes GraphSnapshot.
+    /// Returns a subset of `edges`, consisting of edges which were not in the graph before.
+    pub fn update(
+        &mut self,
+        clock: &time::Clock,
+        mut edges: Vec<Edge>,
+        unreliable_peers: &HashSet<PeerId>,
+    ) -> (Vec<Edge>, GraphSnapshot) {
+        let _next_hops_recalculation = metrics::ROUTING_TABLE_RECALCULATION_HISTOGRAM.start_timer();
+        let total = edges.len();
+        // load the components BEFORE updating the edges.
+        // so that result doesn't contain edges we already have in storage.
+        // It is especially important for initial full sync with peers, because
+        // we broadcast all the returned edges to all connected peers.
+        let now = clock.now_utc();
+        for edge in &edges {
+            let key = edge.key();
+            self.load_component(now, key.0.clone());
+            self.load_component(now, key.1.clone());
+        }
+        edges.retain(|e| self.update_edge(now, e.clone()));
+        // Update metrics after edge update
+        metrics::EDGE_UPDATES.inc_by(total as u64);
+        metrics::EDGE_ACTIVE.set(self.graph.total_active_edges() as i64);
```

Comment:
nit: shouldn't we update them after the prune_old_edges/prune unreachable peers call?

Reply:
makes sense, moved all metric updating to the bottom of the function